graphics/include/ignition/common/SubMesh.hh

```diff
@@ -284,8 +285,15 @@ namespace ignition
       public: void SetMaterialIndex(const unsigned int _index);
 
       /// \brief Get the material index
-      /// \return The assigned material index.
-      public: unsigned int MaterialIndex() const;
+      /// \return The assigned material index. If no material index is assigned
+      /// to this submesh, std::numeric_limits<unsigned int>::max() is returned.
+      /// \note This method is deprecated, use GetMaterialIndex instead
+      public: unsigned int IGN_DEPRECATED(5) MaterialIndex() const;
+
+      /// \brief Get the material index
+      /// \return The assigned material index. Nullopt is returned if the
+      /// submesh has no assigned material index
+      public: std::optional<unsigned int> GetMaterialIndex() const;
```

Comment:
We are not using the prefix `Get` to define the getters

Reply:
I believe we need a new name for the method in order to change the return value, so GetMaterialIndex seems a reasonable choice, though it does break our naming convention.

Reply:
> I believe we need a new name for the method in order to change the return value

Yes, that's right - see https://github.com/ignitionrobotics/ign-common/issues/317#issuecomment-1055965199. This code won't compile if `MaterialIndex` is used since it would result in two `MaterialIndex` methods with different return values, which is ambiguous. If we don't want to use the `Get` prefix here, what should we use instead?

Reply:
As a follow-up to this, a workaround to avoid using the `Get` prefix is to make this new method take a mutable input parameter that the method can write to. This is what's being done for `gazebo::EntityComponentManager::State` to differentiate between the `msgs::SerializedState` and `msgs::SerializedStateMap` types:
* https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo6_6.7.0/include/ignition/gazebo/EntityComponentManager.hh#L554-L564
* https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo6_6.7.0/include/ignition/gazebo/EntityComponentManager.hh#L609-L624

So, for this PR, the new method would look like this:
```c++
MaterialIndex(std::optional<unsigned int> &_idx) const;
```

What do others think about this approach?

Reply:
that also works; either way is fine in my opinion

it's slightly easier to use with a `return` value instead of a mutable input

~~~
audo index = submesh->GetMaterialIndex();
// vs
std::optional<unsigned int> index;
submesh->MaterialIndex(index);
~~~

but our pattern is not to use `Get` in method names

Reply:
> it's slightly easier to use with a return value instead of a mutable input

I agree; okay, I will leave this as-is unless anyone else has another idea. I don't think there's a way to get around the `Get` prefix for this scenario if we decide to take the return value approach (unless someone else has an idea for a different method name).